File: src/main/memory-index.ts
// memory-index.ts
// Three-tier Hebbian memory system for the EA agent.
//
// Architecture (Khoj's TextToEntries base + Hebb 1949 reinforcement):
//
//   Tier 1 — Working Memory (MEMORY.md, always in system prompt, ≤50 lines)
//     Pointers only. Zero loading cost.
//
//   Tier 2 — Indexed Memory (memory/*.md + index.json)
//     One file per topic. Loaded on demand. Scored by weight.
//     weight range: 0.0 – 1.0
//     decay: weight -= decay_rate per day (reset on access)
//     promotion: mentions ≥ 3 → weight = 0.8
//
//   Tier 3 — Archive (memory/archive/YYYY-MM-DD.md)
//     Daily append-only. Loaded only on explicit recall.
//     Entries with weight < 0.05 are pruned here weekly.
//
// MD5 dedup (Khoj pattern): skip if content hash already in index.

import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import { app } from 'electron';
// Zero-tolerance privacy backstop. RULE_NO_FORBIDDEN_PEOPLE is the same
// word-boundary matcher the output critics use; importing it here wires that
// (previously output-only) guard into the durable-memory write boundary so the
// "no ex-wife tracking" rule holds at persistence, matching the already-wired
// EC2 scripts/lib/graphiti-source-policy.js. Pure module, no cycle.
import { RULE_NO_FORBIDDEN_PEOPLE } from './output-critic-rules';
// Graphiti cascade: every Tier 2 write also fires addEpisode() so the
// knowledge graph stays in sync with the filesystem. Statically imported:
// the previous lazy require('./graphiti-client') did not survive bundling
// (no ./graphiti-client chunk exists next to out/main/index.js), which left
// the cascade silently dead in the built app. Add failures are no longer
// dropped either: they spool to graphiti-spool.jsonl and replay on the next
// successful Graphiti interaction (see graphiti-retry-queue.ts).
// Reference: AMY_DEEP_RESEARCH.md section 1.
import {
  addEpisode as graphitiAddEpisode,
  isGraphitiAvailable,
  type GraphitiEpisode,
} from './graphiti-client';
import { drainSpool, enqueueFailedEpisode, spoolCount } from './graphiti-retry-queue';

// Under vitest the cascade is opt-in: unit tests that do not mock
// ./graphiti-client must never fire real network calls at the live knowledge
// graph. Tests that DO mock it set SECONDBRAIN_GRAPHITI_CASCADE_UNDER_TEST=1.
function graphitiCascadeEnabled(): boolean {
  if (!process.env.VITEST) return true;
  return process.env.SECONDBRAIN_GRAPHITI_CASCADE_UNDER_TEST === '1';
}

// ── Types ─────────────────────────────────────────────────────────────────────

export interface MemoryEntry {
  id: string; // MD5 hash of normalized content
  topic: string; // short label, e.g. "dentist-project"
  file: string; // relative path under memory/ (e.g. "Employer-metis.md")
  weight: number; // 0.0 – 1.0 Hebbian weight
  mentions: number; // total access count
  last_accessed: string; // ISO date
  decay_rate: number; // how fast it fades (0.02 = slow, 0.10 = fast)
  valid_at: string; // when this fact became true
  invalid_at?: string; // when this fact was superseded (never delete, just mark)
  tier: 1 | 2 | 3;
}

export interface MemoryIndex {
  version: number;
  last_updated: string;
  entries: MemoryEntry[];
  hashes: string[]; // MD5 set for dedup
}

// ── Path helpers ──────────────────────────────────────────────────────────────

function memoryRoot(): string {
  return path.join(app.getPath('userData'), 'data', 'agent', 'memory');
}

function archiveDir(): string {
  return path.join(memoryRoot(), 'archive');
}

function indexPath(): string {
  return path.join(memoryRoot(), 'index.json');
}

function workingMemoryPath(): string {
  return path.join(memoryRoot(), 'MEMORY.md');
}

function tier2FilePath(file: string): string {
  return path.join(memoryRoot(), file);
}

function archiveFilePath(date: string): string {
  return path.join(archiveDir(), `${date}.md`);
}

// ── Index I/O ─────────────────────────────────────────────────────────────────

let _indexCache: MemoryIndex | null = null;
let _indexCachedAt = 0;
const INDEX_CACHE_TTL = 2 * 60 * 1000; // 2 minutes

export function loadIndex(): MemoryIndex {
  if (_indexCache && Date.now() - _indexCachedAt < INDEX_CACHE_TTL) {
    return _indexCache;
  }

  const p = indexPath();
  if (!fs.existsSync(p)) {
    const fresh: MemoryIndex = { version: 1, last_updated: now(), entries: [], hashes: [] };
    saveIndex(fresh);
    return fresh;
  }

  try {
    _indexCache = JSON.parse(fs.readFileSync(p, 'utf-8')) as MemoryIndex;
    _indexCachedAt = Date.now();
    return _indexCache;
  } catch {
    const fresh: MemoryIndex = { version: 1, last_updated: now(), entries: [], hashes: [] };
    return fresh;
  }
}

function saveIndex(index: MemoryIndex): void {
  const dir = memoryRoot();
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  index.last_updated = now();
  fs.writeFileSync(indexPath(), JSON.stringify(index, null, 2), 'utf-8');
  _indexCache = index;
  _indexCachedAt = Date.now();
}

function now(): string {
  return new Date().toISOString().slice(0, 10);
}

function md5(content: string): string {
  return crypto.createHash('md5').update(content.trim()).digest('hex');
}

// ── Working Memory (Tier 1) ───────────────────────────────────────────────────

const WORKING_MEMORY_MAX_LINES = 50;

export function readWorkingMemory(): string {
  const p = workingMemoryPath();
  if (!fs.existsSync(p)) return '';
  return fs.readFileSync(p, 'utf-8');
}

export function writeWorkingMemory(content: string): void {
  const dir = memoryRoot();
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });

  // Enforce ≤50 lines — trim oldest entries if needed
  const lines = content.split('\n');
  const trimmed =
    lines.length > WORKING_MEMORY_MAX_LINES
      ? lines.slice(lines.length - WORKING_MEMORY_MAX_LINES).join('\n')
      : content;

  fs.writeFileSync(workingMemoryPath(), trimmed, 'utf-8');
}

export function appendWorkingMemory(line: string): void {
  assertNoForbiddenPeople(line);
  const existing = readWorkingMemory();
  const dated = `[${now()}] ${line}`;
  writeWorkingMemory(existing + '\n' + dated);
}

// ── Privacy backstop ──────────────────────────────────────────────────────────

/**
 * Thrown when content destined for durable memory ExampleCos a name under the
 * zero-tolerance privacy rule (Andrea, Julia, Marie-Louise). The write and the
 * Graphiti cascade are both aborted before any side effect.
 */
export class ForbiddenContentError extends Error {
  readonly matches: string[];
  constructor(matches: string[]) {
    super(
      `Refusing to persist memory: forbidden name(s) ${matches.join(', ')} ` +
        `violate the no-ex-wife-tracking privacy rule. Nothing was written and ` +
        `the Graphiti cascade was skipped.`,
    );
    this.name = 'ForbiddenContentError';
    this.matches = matches;
  }
}

/**
 * Backstop run before any durable-memory write. Throws ForbiddenContentError
 * (loud, never silent) if any part contains a forbidden name. Word-boundary
 * aware via RULE_NO_FORBIDDEN_PEOPLE, so "Andreas"/"Juliana" pass clean.
 */
function assertNoForbiddenPeople(...parts: string[]): void {
  const issues = RULE_NO_FORBIDDEN_PEOPLE.check(parts.join('\n'));
  if (issues && issues.length) {
    const matches = Array.from(new Set(issues.map((i) => i.match).filter((m): m is string => !!m)));
    const found = matches.length ? matches : ['(redacted)'];
    console.error(`[memory-index] ${new ForbiddenContentError(found).message}`);
    throw new ForbiddenContentError(found);
  }
}

// ── Tier 2: Indexed Memory ────────────────────────────────────────────────────

/**
 * Add or update a memory entry. MD5 dedup — if content is identical, just
 * bumps the mention count and resets the decay clock.
 */
export function upsertMemory(
  topic: string,
  content: string,
  opts?: { decayRate?: number; file?: string },
): MemoryEntry {
  assertNoForbiddenPeople(topic, content);
  const index = loadIndex();
  const hash = md5(content);

  // Dedup check
  const existingEntry = index.entries.find((e) => e.id === hash && !e.invalid_at);
  if (existingEntry) {
    existingEntry.mentions++;
    existingEntry.last_accessed = now();
    existingEntry.decay_rate = existingEntry.mentions >= 3 ? 0.02 : 0.1;
    if (existingEntry.mentions >= 3 && existingEntry.weight < 0.5) {
      existingEntry.weight = 0.8; // promotion
    }
    existingEntry.weight = Math.min(1.0, existingEntry.weight);
    saveIndex(index);
    return existingEntry;
  }

  // New entry
  const fileName = opts?.file ?? slugify(topic) + '.md';
  const entry: MemoryEntry = {
    id: hash,
    topic,
    file: fileName,
    weight: 0.2,
    mentions: 1,
    last_accessed: now(),
    decay_rate: opts?.decayRate ?? 0.1,
    valid_at: now(),
    tier: 2,
  };

  // Write content to tier 2 file
  const filePath = tier2FilePath(fileName);
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });

  const header = `# ${topic}\n*weight: ${entry.weight} | mentions: ${entry.mentions} | valid_from: ${entry.valid_at}*\n\n`;
  fs.writeFileSync(filePath, header + content, 'utf-8');

  index.entries.push(entry);
  index.hashes.push(hash);
  saveIndex(index);

  // Graphiti cascade: non-blocking and never throws into the write path.
  // A failed add is spooled instead of silently lost; a successful add
  // opportunistically drains the spool (the tunnel is evidently up).
  if (graphitiCascadeEnabled()) {
    void fireGraphitiCascade({
      name: `tier2:${topic}`,
      episode_body: content,
      source_description: `memory_upsert:tier2:${topic}`,
      reference_time: entry.valid_at,
      group_id: 'owner-ea',
    });
  }

  return entry;
}

async function fireGraphitiCascade(episode: GraphitiEpisode): Promise<void> {
  let accepted = false;
  try {
    accepted = await graphitiAddEpisode(episode);
  } catch {
    accepted = false;
  }
  try {
    if (accepted) {
      await maybeDrainSpool();
    } else {
      enqueueFailedEpisode(episode);
    }
  } catch {
    /* spool I/O is best-effort; the Tier 2 write itself already succeeded */
  }
}

/**
 * Replay episodes spooled while Graphiti was unreachable. Called after a
 * successful Graphiti add (tunnel known up) and from initMemoryIndex when
 * the spool is non-empty and Graphiti reports healthy. Never throws.
 */
export async function maybeDrainSpool(): Promise<{
  attempted: number;
  sent: number;
  remaining: number;
} | null> {
  try {
    if (spoolCount() === 0) return null;
    return await drainSpool((ep) => graphitiAddEpisode(ep));
  } catch {
    return null;
  }
}

/**
 * Mark an existing memory entry as superseded (never deleted, just flagged).
 * Optionally provide the replacement entry.
 */
export function invalidateMemory(id: string, replacementContent?: string): void {
  const index = loadIndex();
  const entry = index.entries.find((e) => e.id === id);
  if (entry) {
    entry.invalid_at = now();
    entry.weight = 0;
  }
  saveIndex(index);

  if (replacementContent && entry) {
    upsertMemory(entry.topic, replacementContent);
  }
}

/**
 * Load all Tier 2 entries with weight ≥ threshold (default 0.3).
 * Returns their full content for system prompt injection.
 */
export function loadRelevantMemories(
  minWeight = 0.3,
  maxEntries = 8,
): Array<{ topic: string; content: string; weight: number }> {
  const index = loadIndex();
  const relevant = index.entries
    .filter((e) => e.tier === 2 && !e.invalid_at && e.weight >= minWeight)
    .sort((a, b) => b.weight - a.weight)
    .slice(0, maxEntries);

  const results: Array<{ topic: string; content: string; weight: number }> = [];
  for (const entry of relevant) {
    const filePath = tier2FilePath(entry.file);
    if (!fs.existsSync(filePath)) continue;
    try {
      const content = fs.readFileSync(filePath, 'utf-8');
      results.push({ topic: entry.topic, content, weight: entry.weight });

      // Bump mention count (access = reinforcement)
      entry.mentions++;
      entry.last_accessed = now();
    } catch {
      /* skip unreadable files */
    }
  }

  if (results.length > 0) saveIndex(index);
  return results;
}

// ── Tier 3: Archive ───────────────────────────────────────────────────────────

/** Append a fact or summary to today's archive file (append-only). */
export function appendToArchive(content: string): void {
  const dir = archiveDir();
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });

  const filePath = archiveFilePath(now());
  const entry = `\n---\n*${new Date().toISOString()}*\n${content}\n`;
  fs.appendFileSync(filePath, entry, 'utf-8');
}

/** Load a specific archive date for explicit recall. */
export function loadArchiveDate(date: string): string {
  const filePath = archiveFilePath(date);
  if (!fs.existsSync(filePath)) return '';
  return fs.readFileSync(filePath, 'utf-8');
}

// ── Nightly decay & pruning ───────────────────────────────────────────────────

/**
 * Apply daily decay to all Tier 2 entries.
 * Entries that decay below 0.05 are moved to archive.
 * Run this once per night (acquireLock before calling).
 */
export function runNightlyDecay(): { decayed: number; archived: number; pruned: number } {
  const index = loadIndex();
  let decayed = 0;
  let archived = 0;
  let pruned = 0;

  for (const entry of index.entries) {
    if (entry.tier !== 2 || entry.invalid_at) continue;

    // Apply decay
    const daysSinceAccess = Math.floor(
      (Date.now() - new Date(entry.last_accessed).getTime()) / (1000 * 60 * 60 * 24),
    );

    if (daysSinceAccess > 0) {
      const oldWeight = entry.weight;
      // Multiplicative decay: weight *= (1 - decay_rate) per day
      // Much gentler than subtractive — a new entry (0.2, rate 0.10)
      // lasts ~15 days vs. 2 days with the old formula.
      entry.weight = Math.max(0, entry.weight * Math.pow(1 - entry.decay_rate, daysSinceAccess));
      if (entry.weight !== oldWeight) decayed++;
    }

    // Archive entries below threshold
    if (entry.weight < 0.05 && !entry.invalid_at) {
      const filePath = tier2FilePath(entry.file);
      if (fs.existsSync(filePath)) {
        const content = fs.readFileSync(filePath, 'utf-8');
        appendToArchive(
          `## ${entry.topic} (archived — weight: ${entry.weight.toFixed(3)})\n${content}`,
        );
        fs.unlinkSync(filePath);
        archived++;
      }
      // Remove from index (it's in archive now)
      index.entries = index.entries.filter((e) => e.id !== entry.id);
      index.hashes = index.hashes.filter((h) => h !== entry.id);
      pruned++;
    }
  }

  saveIndex(index);
  return { decayed, archived, pruned };
}

/**
 * Build a concise memory context string for system prompt injection.
 * Tier 1 (working memory) always included. Tier 2 loaded by weight.
 */
export function buildMemoryContext(opts?: { maxChars?: number; minWeight?: number }): string {
  const maxChars = opts?.maxChars ?? 3000;
  const minWeight = opts?.minWeight ?? 0.3;

  const working = readWorkingMemory();
  const tier2 = loadRelevantMemories(minWeight, 6);

  const parts: string[] = [];

  if (working.trim()) {
    parts.push(`### Working Memory\n${working.trim()}`);
  }

  for (const m of tier2) {
    parts.push(`### ${m.topic} (weight: ${m.weight.toFixed(2)})\n${m.content.slice(0, 500)}`);
  }

  const combined = parts.join('\n\n');
  return combined.length > maxChars
    ? combined.slice(0, maxChars) + '\n\n*(memory truncated)*'
    : combined;
}

// ── Helpers ───────────────────────────────────────────────────────────────────

function slugify(text: string): string {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 50);
}

// ── Init ──────────────────────────────────────────────────────────────────────

export function initMemoryIndex(): void {
  const dir = memoryRoot();
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
  const archDir = archiveDir();
  if (!fs.existsSync(archDir)) fs.mkdirSync(archDir, { recursive: true });
  loadIndex(); // ensures index.json exists
  console.log(`[memory-index] Initialized at ${dir}`);
  // Replay any Graphiti episodes spooled while the SSH tunnel was down.
  if (graphitiCascadeEnabled()) {
    void (async () => {
      try {
        if (spoolCount() > 0 && (await isGraphitiAvailable())) {
          await maybeDrainSpool();
        }
      } catch {
        /* best-effort replay; the next successful add drains too */
      }
    })();
  }
}

File: python/memory_index.py
"""
Three-tier Hebbian memory store.

Port of secondbrain-os/src/main/memory-index.ts. The Graphiti cascade is
intentionally omitted; the TS version treats it as fire-and-forget anyway.

Tier 1  MEMORY.md         always-loaded pointer file, capped at 50 lines
Tier 2  <topic>.md + index.json   one file per topic, weighted, decays
Tier 3  archive/YYYY-MM-DD.md     append-only, entries land here below 0.05

Weight rules:
  new entry          weight 0.2, decay_rate 0.1
  re-access (dedup)  mentions++, last_accessed reset
  promotion          mentions >= 3 -> weight 0.8, decay_rate 0.02
  daily decay        weight *= (1 - decay_rate) ** days_since_access
  archive threshold  weight < 0.05  -> moved to archive, removed from index
"""
from __future__ import annotations

import hashlib
import json
import re
from dataclasses import asdict, dataclass, field
from datetime import date, datetime, timedelta, timezone
from pathlib import Path
from typing import Optional

WORKING_MEMORY_MAX_LINES = 50
ARCHIVE_THRESHOLD = 0.05
PROMOTION_MENTIONS = 3
PROMOTION_WEIGHT = 0.8
INITIAL_WEIGHT = 0.2
INITIAL_DECAY = 0.1
PROMOTED_DECAY = 0.02


@dataclass
class MemoryEntry:
    id: str
    topic: str
    file: str
    weight: float
    mentions: int
    last_accessed: str
    decay_rate: float
    valid_at: str
    tier: int = 2
    invalid_at: Optional[str] = None


@dataclass
class MemoryIndex:
    version: int = 1
    last_updated: str = field(default_factory=lambda: date.today().isoformat())
    entries: list[MemoryEntry] = field(default_factory=list)
    hashes: list[str] = field(default_factory=list)


def _today() -> str:
    return date.today().isoformat()


def _md5(content: str) -> str:
    return hashlib.md5(content.strip().encode("utf-8")).hexdigest()


def _slugify(text: str) -> str:
    return re.sub(r"[^a-z0-9]+", "-", text.lower()).strip("-")[:50]


class MemoryStore:
    """Three-tier memory store rooted at a directory on disk."""

    def __init__(self, root: Path) -> None:
        self.root = Path(root)
        self.archive_dir = self.root / "archive"
        self.index_path = self.root / "index.json"
        self.working_path = self.root / "MEMORY.md"
        self._cache: Optional[MemoryIndex] = None

    def init(self) -> None:
        self.root.mkdir(parents=True, exist_ok=True)
        self.archive_dir.mkdir(parents=True, exist_ok=True)
        self._load_index()

    # ── Index I/O ──────────────────────────────────────────────────────────
    def _load_index(self) -> MemoryIndex:
        if self._cache is not None:
            return self._cache
        if not self.index_path.exists():
            self._cache = MemoryIndex()
            self._save_index()
            return self._cache
        try:
            raw = json.loads(self.index_path.read_text("utf-8"))
            self._cache = MemoryIndex(
                version=raw.get("version", 1),
                last_updated=raw.get("last_updated", _today()),
                entries=[MemoryEntry(**e) for e in raw.get("entries", [])],
                hashes=list(raw.get("hashes", [])),
            )
            return self._cache
        except (json.JSONDecodeError, TypeError, KeyError):
            self._cache = MemoryIndex()
            return self._cache

    def _save_index(self) -> None:
        assert self._cache is not None
        self._cache.last_updated = _today()
        self.root.mkdir(parents=True, exist_ok=True)
        payload = {
            "version": self._cache.version,
            "last_updated": self._cache.last_updated,
            "entries": [asdict(e) for e in self._cache.entries],
            "hashes": list(self._cache.hashes),
        }
        self.index_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")

    # ── Tier 1: Working Memory ─────────────────────────────────────────────
    def read_working(self) -> str:
        if not self.working_path.exists():
            return ""
        return self.working_path.read_text("utf-8")

    def write_working(self, content: str) -> None:
        self.root.mkdir(parents=True, exist_ok=True)
        lines = content.split("\n")
        if len(lines) > WORKING_MEMORY_MAX_LINES:
            lines = lines[-WORKING_MEMORY_MAX_LINES:]
        self.working_path.write_text("\n".join(lines), encoding="utf-8")

    def append_working(self, line: str) -> None:
        existing = self.read_working()
        dated = f"[{_today()}] {line}"
        self.write_working(f"{existing}\n{dated}" if existing else dated)

    # ── Tier 2: Indexed Memory ─────────────────────────────────────────────
    def upsert(
        self,
        topic: str,
        content: str,
        *,
        decay_rate: Optional[float] = None,
        file: Optional[str] = None,
    ) -> MemoryEntry:
        index = self._load_index()
        h = _md5(content)

        existing = next(
            (e for e in index.entries if e.id == h and e.invalid_at is None),
            None,
        )
        if existing is not None:
            existing.mentions += 1
            existing.last_accessed = _today()
            # Mirrors TS quirk: any per-entry custom rate is overwritten on re-access.
            existing.decay_rate = (
                PROMOTED_DECAY if existing.mentions >= PROMOTION_MENTIONS else INITIAL_DECAY
            )
            if existing.mentions >= PROMOTION_MENTIONS and existing.weight < 0.5:
                existing.weight = PROMOTION_WEIGHT
            existing.weight = min(1.0, existing.weight)
            self._save_index()
            return existing

        fname = file or f"{_slugify(topic)}.md"
        entry = MemoryEntry(
            id=h,
            topic=topic,
            file=fname,
            weight=INITIAL_WEIGHT,
            mentions=1,
            last_accessed=_today(),
            decay_rate=INITIAL_DECAY if decay_rate is None else decay_rate,
            valid_at=_today(),
            tier=2,
        )

        file_path = self.root / fname
        file_path.parent.mkdir(parents=True, exist_ok=True)
        header = (
            f"# {topic}\n"
            f"*weight: {entry.weight} | mentions: {entry.mentions} "
            f"| valid_from: {entry.valid_at}*\n\n"
        )
        file_path.write_text(header + content, encoding="utf-8")

        index.entries.append(entry)
        index.hashes.append(h)
        self._save_index()
        return entry

    def invalidate(
        self, entry_id: str, replacement_content: Optional[str] = None
    ) -> None:
        index = self._load_index()
        target = next((e for e in index.entries if e.id == entry_id), None)
        if target is not None:
            target.invalid_at = _today()
            target.weight = 0.0
        self._save_index()
        if replacement_content and target is not None:
            self.upsert(target.topic, replacement_content)

    def load_relevant(
        self, min_weight: float = 0.3, max_entries: int = 8
    ) -> list[tuple[MemoryEntry, str]]:
        index = self._load_index()
        relevant = sorted(
            (
                e
                for e in index.entries
                if e.tier == 2 and e.invalid_at is None and e.weight >= min_weight
            ),
            key=lambda e: e.weight,
            reverse=True,
        )[:max_entries]

        results: list[tuple[MemoryEntry, str]] = []
        touched = False
        for entry in relevant:
            file_path = self.root / entry.file
            if not file_path.exists():
                continue
            results.append((entry, file_path.read_text("utf-8")))
            entry.mentions += 1
            entry.last_accessed = _today()
            touched = True
        if touched:
            self._save_index()
        return results

    # ── Tier 3: Archive ────────────────────────────────────────────────────
    def append_archive(self, content: str) -> None:
        self.archive_dir.mkdir(parents=True, exist_ok=True)
        file_path = self.archive_dir / f"{_today()}.md"
        stamp = datetime.now(timezone.utc).isoformat()
        with file_path.open("a", encoding="utf-8") as f:
            f.write(f"\n---\n*{stamp}*\n{content}\n")

    def load_archive(self, day: str) -> str:
        file_path = self.archive_dir / f"{day}.md"
        if not file_path.exists():
            return ""
        return file_path.read_text("utf-8")

    # ── Nightly decay ──────────────────────────────────────────────────────
    def run_nightly_decay(self) -> dict[str, int]:
        index = self._load_index()
        decayed = 0
        archived = 0
        pruned = 0
        to_remove: set[str] = set()

        for entry in list(index.entries):
            if entry.tier != 2 or entry.invalid_at is not None:
                continue

            days = (date.today() - date.fromisoformat(entry.last_accessed)).days
            if days > 0:
                old = entry.weight
                entry.weight = max(
                    0.0, entry.weight * ((1.0 - entry.decay_rate) ** days)
                )
                if entry.weight != old:
                    decayed += 1

            if entry.weight < ARCHIVE_THRESHOLD and entry.invalid_at is None:
                file_path = self.root / entry.file
                if file_path.exists():
                    self.append_archive(
                        f"## {entry.topic} (archived — weight: {entry.weight:.3f})\n"
                        f"{file_path.read_text('utf-8')}"
                    )
                    file_path.unlink()
                    archived += 1
                to_remove.add(entry.id)
                pruned += 1

        if to_remove:
            index.entries = [e for e in index.entries if e.id not in to_remove]
            index.hashes = [h for h in index.hashes if h not in to_remove]

        self._save_index()
        return {"decayed": decayed, "archived": archived, "pruned": pruned}

    # ── Context builder ────────────────────────────────────────────────────
    def build_context(self, *, max_chars: int = 3000, min_weight: float = 0.3) -> str:
        working = self.read_working()
        tier2 = self.load_relevant(min_weight=min_weight, max_entries=6)

        parts: list[str] = []
        if working.strip():
            parts.append(f"### Working Memory\n{working.strip()}")
        for entry, content in tier2:
            parts.append(
                f"### {entry.topic} (weight: {entry.weight:.2f})\n{content[:500]}"
            )

        combined = "\n\n".join(parts)
        if len(combined) > max_chars:
            return combined[:max_chars] + "\n\n*(memory truncated)*"
        return combined


if __name__ == "__main__":
    import tempfile

    with tempfile.TemporaryDirectory() as tmp:
        store = MemoryStore(Path(tmp) / "memory")
        store.init()
        store.write_working("Owner: Demo User\nTimezone: CT")

        for _ in range(3):
            store.upsert("dentist project", "Found a dentist who skips X-rays.")
        store.upsert("low-stakes note", "Tried a new coffee shop.")

        # Simulate 30 days of inactivity on the low-stakes note.
        idx = store._load_index()
        for e in idx.entries:
            if e.topic == "low-stakes note":
                e.last_accessed = (date.today() - timedelta(days=30)).isoformat()
        store._save_index()

        print("Before decay:")
        for e in store._load_index().entries:
            print(f"  {e.topic:25s} w={e.weight:.3f} m={e.mentions}")

        stats = store.run_nightly_decay()
        print(f"\nDecay stats: {stats}\n")

        print("After decay:")
        for e in store._load_index().entries:
            print(f"  {e.topic:25s} w={e.weight:.3f} m={e.mentions}")

        print("\n--- build_context() ---")
        print(store.build_context())
